[1.19.x] Add system to allow mods to change when ItemEntities should be picked up

ApexModder opened the pull request:
This PR adds a new method into IForgeItem which modders should override to dictate when / if their custom items should be able to be picked up or not.

By default this method checks to see if the legacy `PreventRemoteMovement` tags exist on the ItemEntity or not, afaik this tag has commonly been used to disable magnet and other remote item movement logic, which this new method is set to replace.
_Legacy check purely exists for backwards compatibility._

Implementations should override this method and check if the ItemStack / ItemEntity is in any state where they would not want objects like Hoppers, Players, Create-Mod belts etc to pick up the item.

Consumers should use this method to check if the ItemStack / ItemEntity can currently be picked up.

The method takes 3 parameters
- ItemStack, The ItemStack object being picked up [_pulled from the ItemEntity_].
- ItemEntity, The ItemEntity object being picked up.
- boolean isRemoteMovement, flag to say what kind of pickup is currently being processed.
	- This should be true and only ever true for remote movement pickup types.
	- This could be Magnet Items, Hoppers, Vacuum Chests, Create-Mod belts, etc. Really anything other than a LivingEntity touching the ItemEntity.

	- This should be false and only ever false when a LivingEntity is trying to pick up the ItemEntity via touching it.
	- This could be a regular Player pickup, Zombies picking up loot, or any other custom mob trying to pick up the loot.

2 handly little utility/helper methods are included to easier call this method from other means.
- `IForgeItemStack#canBePickedUp(ItemEntity, boolean)`
- `ItemEntity#canBePickedUp()`
These methods simply just wrap all the way down to `IForgeItem#canBePickedUp(ItemStack, ItemEntity, boolean)`.


I have included a little test mod that registers 1 new item to the game, this item disallows being picked up by remote means while still allowing regular pickup methods.
This means remote sources [_Hoppers, Magents, etc_] will not pick up the item, while Players & other LivingEntities will.

eerussianguy commented on src/main/java/net/minecraftforge/common/extensions/IForgeItem.java:
```diff
@@ -845,4 +845,31 @@ default boolean isNotReplaceableByPickAction(ItemStack stack, Player player, int
+     */
+    default boolean canBePickedUp(@NotNull ItemStack stack, @NotNull ItemEntity entity, boolean isRemoteMovement)
+    {
+        if(isRemoteMovement) {
```
braces on newlines

eerussianguy commented on patches/minecraft/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch:
```diff
@@ -28,6 +28,14 @@
+    }
+ 
+    public static boolean m_59331_(Container p_59332_, ItemEntity p_59333_) {
++      if(!p_59333_.canBePickedUp(true)) return false; // Apex: Check to see if the item currently disallows being picked up
```
why are your patches signed as Apex? Typically forge patches say `FORGE:`

eerussianguy commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -0,0 +1,50 @@
+
+	public ItemPreventPickupTest()
+	{
+		if(ENABLE)
+			ITEMS.register(FMLJavaModLoadingContext.get().getModEventBus());
```
This is a weird indent

ChampionAsh5357 reviewed (changes requested):
Just a general heads up, you need to fix the comments (Forge, not Apex), LF at EOF, 4 spaces instead of tabs, and a bit of documentation on the event.

eerussianguy commented on src/main/java/net/minecraftforge/common/extensions/IForgeItemStack.java:
```diff
@@ -572,4 +572,19 @@ default boolean isNotReplaceableByPickAction(Player player, int inventorySlot)
+     * @see IForgeItem#canBePickedUp(ItemStack, ItemEntity, boolean)
+     * @param entity the item entity being picked up
+     * @param isRemoteMovement true if pickup was caused by remote movement or not
+     * @return true if this item should be picked up, false otherwise
```
`{@return if this item should be picked up}`

TheCurle commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -0,0 +1,50 @@
+package net.minecraftforge.debug.item;
```
Forge classes should have license headers.

ApexModder commented:
Sorry for the various issues with the code, this is my first pr into the forge code base & vanilla patches, and wasn't aware of some of the stuff you all mentioned. Hopefully the commits I just pushed resolved them correctly.

Edit:
> Just a general heads up, you need to fix the comments (Forge, not Apex), LF at EOF, 4 spaces instead of tabs, and a bit of documentation on the event.

tbh I was not aware that my ide was doing this, think I changed the correct setting to keep line endings and blank lines at end of files.

one thing I did make sure to disable though, was the automatic import optimizations I usually have enabled, so hopefully, those should be fine and as they were.

eerussianguy commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -268,6 +268,8 @@ modId="custom_item_decorations_test"
+[[mods]]
+modId="item_prevent_pickup_test"
 
 
-# ADD ABOVE THIS LINE
+# ADD ABOVE THIS LINE
```
whitespace change

ApexModder replied:
Ah I missed a file that had a new line removed, I will add that back real quick

Random832 commented:
I don't know if this really covers all the use cases. Like, for example, for Botania, the prevent remote movement flag specifically applies to magnets but *not* to the hopperhock [a flower that picks up item entities in an area and depots them in adjacent containers].

Putting the item in charge of the logic and declaring the PreventRemoteMovement tag to be a "legacy" also discounts the scenario where you want to have a device that prevents magnets from working on all items in an area, or e.g. Immersive Engineering wants to prevent magnets from working on items that are on top of conveyor belts while obviously allowing the conveyor belts themselves to move the item.

I do think it'd be a good idea to look at providing more nuance than a boolean PreventRemoteMovement flag, but this requires more design work to support more than just your particular use case (and as several people have pointed out on discord, the entity in your particular use case probably should not be an item entity at all).

Maybe an event would be more suitable than an IForgeItem method?

I also recall a discussion earlier in the week about preventing non-player entities from picking up items, including the (optional) LivingEntity in the API could allow it to be used for this. Perhaps we could also use an enum for broad categories of remote movement (player magnets, things like fans or vector plates, things that put immediately touching items in a container, things that put items in an AOE in a container) rather than a simple true/false - or maybe even a 'registry' (or at least namespaced strings of some kind) to identify the specific mover responsible for trying to pick up an item.

pupnewfster reviewed (commented):
I wonder if it would make more sense to patch the logic into the `ItemEntity` so that the `ItemEntity`'s existing preventing pickup logic is what gets used and the item just gets queried if it should prevent it by default. As `ItemEntity`'s have a `setNeverPickUp` method on them which vanilla uses to prevent items from ever being picked up (for purposes of spawning items in adventure maps that can't be picked up). For one thing the patched in method `ItemEntity#canBePickedUp` entirely ignores whether the vanilla and the item entity thinks it can be picked up...

ApexModder commented:
> I don't know if this really covers all the use cases. Like, for example, for Botania, the prevent remote movement flag specifically applies to magnets but _not_ to the hopperhock [a flower that picks up item entities in an area and depots them in adjacent containers].
> 
> Putting the item in charge of the logic and declaring the PreventRemoteMovement tag to be a "legacy" also discounts the scenario where you want to have a device that prevents magnets from working on all items in an area, or e.g. Immersive Engineering wants to prevent magnets from working on items that are on top of conveyor belts while obviously allowing the conveyor belts themselves to move the item.
> 
> I do think it'd be a good idea to look at providing more nuance than a boolean PreventRemoteMovement flag, but this requires more design work to support more than just your particular use case (and as several people have pointed out on discord, the entity in your particular use case probably should not be an item entity at all).
> 
> Maybe an event would be more suitable than an IForgeItem method?
> 
> I also recall a discussion earlier in the week about preventing non-player entities from picking up items, including the (optional) LivingEntity in the API could allow it to be used for this. Perhaps we could also use an enum for broad categories of remote movement (player magnets, things like fans or vector plates, things that put immediately touching items in a container, things that put items in an AOE in a container) rather than a simple true/false - or maybe even a 'registry' (or at least namespaced strings of some kind) to identify the specific mover responsible for trying to pick up an item.

I was originally planning on making it an enum rather than the boolean to specify different pickup states or even an interface that modders would implement for custom pick state/reasons and forge provides basic implementation via an enum for say entity pickup/hopper pickup
but thought a simple boolean would be better as its either being picked up by an entity or "magnet" but yeah something to specify an exact pickup reason would maybe be better than a simple "did entity pickup or magnet-like" boolean flag.


> I wonder if it would make more sense to patch the logic into the `ItemEntity` so that the `ItemEntity`'s existing preventing pickup logic is what gets used and the item just gets queried if it should prevent it by default. As `ItemEntity`'s have a `setNeverPickUp` method on them which vanilla uses to prevent items from ever being picked up (for purposes of spawning items in adventure maps that can't be picked up). For one thing the patched in method `ItemEntity#canBePickedUp` entirely ignores whether the vanilla and the item entity thinks it can be picked up...

I put the Item object in charge of this as I thought that would be the best place for this as its something that is really per ItemStack, having the check in the Item object is an easy place for modders to override and disable for only their specific item but I suppose keeping the ItemEntity in charge of this logic would make more sense as could then disable pickup for not only your item but vanilla items & potentially modded ones to via some event or other means to set the flag on an entity by entity basis.


You both make good arguments and this maybe needs better design work on it than I can come up with, I just wanted a simple way to disable my item entity from being picked up in all means, not just what vanilla allows which is entity/player pickup and was kinda surprised with all the magnet-type blocks/items that pickup items that there wasn't already some method in place to disable this. well other than the nbt bool on the entity item persistent data, which I didn't know of until started making this pr.

pupnewfster commented:
The other part of what I meant is that if the ItemEntity controls it (especially given that is where vanilla has it control it) that means that in theory you shouldn't need patches in the entities themselves for if they can pickup a specific item entity.

ApexModder commented:
yeah now that I have given it some thought and read what you all have said a few times over, this needs to be better designed rather than my quick rough go at it and be fully controlled via the `ItemEntity` maybe potentially also a Forge Event to allow mods to hook and modify the pickable state as a last-ditch attempt.

my initial thoughts on how a better-designed system would go:
- the vanilla `ItemEntity#setNeverPickUp` method should disable the entity from being picked up from any means whatsoever, not just `LivingEntity` types touching to collect it.
- patch in some new methods onto `ItemEntity` which allows disabling/enabling specific pickup reasons.
	patch in additional little helper methods to check if a given pickup reason is disabled/enabled & if pickup is fully disabled, also perhaps?
	_fully disabled if either `#setNeverPickUp` was called or all pickup reasons are disabled_
- add in some system for mods to specify new "PickupReasons".
	I'm thinking this could be something simple like an interface for mods to implement and for Forge to provide an enum that implements said interface for basic pickup reasons.
	
	there perhaps also should be some way of querying the world, and using this query as a "PickupReason"?
	for radius checking and other stately based reasons that may change if the item should be picked up or not.
```java
interface ItemPickupReason { }
enum StandardItemPickupReasons implements ItemPickupReason
{
	// maybe this should be some "ENTITY" reason and have entity type passed [somehow]
	// to allow disallowing on an entity-by-entity basis?
	PLAYER, LIVING_ENTITY,

	//hopper-like block item pickups
	HOPPER,

	// generic magic type reason
	// mods can always add their own custom reason if this doesn't fit
	MAGNET_LIKE
}
```

- when picking/collecting the `ItemEntity`, in order to check if the pickup state is valid, the order would be like so:
	1. has the vanilla `#setNeverPickUp` method been called? if so, the item flat out should not be picked up under any circumstances
	2. check if the specified pickup reason has been disabled, if so, do not allow pickup
		_or if the reason is some form of query and if the query results in false_
	3. check some forge event to allow mods to have a last-ditch attempt to change the pickup state.
	
	if none of the above checks are triggered, the item should be allowed to be picked up.

ApexModder commented:
> yeah now that I have given it some thought and read what you all have said a few times over, this needs to be better designed rather than my quick rough go at it and be fully controlled via the `ItemEntity` maybe potentially also a Forge Event to allow mods to hook and modify the pickable state as a last-ditch attempt.
> 
> my initial thoughts on how a better-designed system would go:
> 
>     * the vanilla `ItemEntity#setNeverPickUp` method should disable the entity from being picked up from any means whatsoever, not just `LivingEntity` types touching to collect it.
> 
>     * patch in some new methods onto `ItemEntity` which allows disabling/enabling specific pickup reasons.
>       patch in additional little helper methods to check if a given pickup reason is disabled/enabled & if pickup is fully disabled, also perhaps?
>       _fully disabled if either `#setNeverPickUp` was called or all pickup reasons are disabled_
> 
>     * add in some system for mods to specify new "PickupReasons".
>       I'm thinking this could be something simple like an interface for mods to implement and for Forge to provide an enum that implements said interface for basic pickup reasons.
>       there perhaps also should be some way of querying the world, and using this query as a "PickupReason"?
>       for radius checking and other stately based reasons that may change if the item should be picked up or not.
> 
> 
> ```java
> interface ItemPickupReason { }
> enum StandardItemPickupReasons implements ItemPickupReason
> {
> 	// maybe this should be some "ENTITY" reason and have entity type passed [somehow]
> 	// to allow disallowing on an entity-by-entity basis?
> 	PLAYER, LIVING_ENTITY,
> 
> 	//hopper-like block item pickups
> 	HOPPER,
> 
> 	// generic magic type reason
> 	// mods can always add their own custom reason if this doesn't fit
> 	MAGNET_LIKE
> }
> ```
> 
>     * when picking/collecting the `ItemEntity`, in order to check if the pickup state is valid, the order would be like so:
>       
>       1. has the vanilla `#setNeverPickUp` method been called? if so, the item flat out should not be picked up under any circumstances
>       2. check if the specified pickup reason has been disabled, if so, do not allow pickup
>          _or if the reason is some form of query and if the query results in false_
>       3. check some forge event to allow mods to have a last-ditch attempt to change the pickup state.
>       
>       if none of the above checks are triggered, the item should be allowed to be picked up.

in the latest commits, I have reverted all previous changes and added a new system that attempted to implement the above system.

Added 2 new ways which work in tandem to try and dictate when an `ItemEntity` is allowed to be picked up.

**Reasons**
`ItemPickupReason`
This class is a simple generic reason as to what & why the item is being picked up and should be used as a general broad term.
want to disable hoppers picking up, disallow the hopper pickup reason.
want to stop players picking up, disallow the player reason.

`ItemPickupReasons` contains forge-provided reasons but the system is not limited to only these reasons.
much like the tool action system using `ItemPickupReason#getOrCreate(String)` modders can easily add a brand new pickup reason, ofc for this new reason to be fully utilized said modders would have to "pickup" the item themselves specifying their new reason as to why its being picked up.

You can allow/disallow reasons using the `ItemEntity#enablePickupReason` & `ItemEntity#disablePickupReason` methods.

**Queries**
`ItemPickupQuery`
This class is used as a predicate to be called and validated against when the item is attempted to be picked up.
These queries can be used to validate against anything in world, on the item entity or even item stack.
The object attempting to pick up the item is also passed across to this query unfortunately as a java `Object` as this could be of any type (Item/Block/BlockEntity/Entity etc) so instance check and cast accordingly.

You can add/remove these queries using `ItemEntity#addPickupQuery` & `ItemEntity#removePickupQuery`

**Validating**
Before modders allow their objects from picking up items they should call the `ItemEntity#isPickupAllowed(Object, ItemPickupReason...)` method, to validate if the item has been disallowed from being picked up or not.

This method takes 2 parameters, the object attempting to pick up the item (entity, magnet/item, hopper/block entity, etc) and any ItemPickupReason as to why this item is being picked up.

We also check against the vanilla way of disabling item pickup, (if `ItemEntity#setNeverPickUp` has been called).
If the vanilla method has been called it forcibly disables the item from being picked up by all means.

**Event**
There is also a new event `ItemPickupAllowedEvent`, this event is used to forcibly allow or disallow if an item can be picked up or not.

This is done using the events result type:
- ALLOW - forcibly allow item to be picked up, ignoring all reasons and queries
- DENY - forcibly disallow the item from being picked up, ignoring all reasons & queries
- DEFAULT - let default logic happen, validate that given reasons are allowed and all provided queries pass successfully

**TestMod**
Provided test mod now does the following
- Adds a new "dummy" item to the game, this item can only be picked up under the following circumstances
	- For players: if sneaking
	- For zombies: if standing above stone
	- For hoppers: if stacked count >= 2
- forcibly disallows iron ingots from ever being picked up (using the new event)
- disallows diamonds from being picked up by hoppers (using the hopper ItemPickupReason)

all of the above are added as tooltips to each of the items to easily see how they can be picked up in-game

ApexModder committed:
Re-implement PR #9112 onto 1.19.3
_Fixes issues with ForgeGradle failing to apply patches_

ApexModder committed:
Fix license header issue [_Add whitespace after license header_]

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/item/ItemPickupAllowedEvent.java:
```diff
@@ -0,0 +1,69 @@
+    private final Object collector;
+    private final ItemPickupReason[] pickupReasons;
+
+    public ItemPickupAllowedEvent(ItemEntity itemEntity, Object collector, ItemPickupReason... pickupReasons)
```
Please mark this with `@ApiStatus.Internal`.

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/item/ItemPickupAllowedEvent.java:
```diff
@@ -0,0 +1,69 @@
+import net.minecraftforge.items.pickup.ItemPickupReason;
+
+/**
+ * Event fired to check if the given item is currently allowed to be picked up or not
+ *
+ * <p>This event is not {@link Cancelable cancellable}, and {@link HasResult has a result}.</p>
+ * <ul>
+ *     <li>{@link Result#ALLOW} - forcibly allow this item to be picked up</li>
+ *     <li>{@link Result#DEFAULT} - let the default logic dictate if this item should be picked up or not, this relies on the {@link ItemEntity#isPickupAllowed(Object, ItemPickupReason...)} method</li>
+ *     <li>{@link Result#DENY} - forcibly disallow this item to be picked up</li>
+ * </ul>
+ *
+ * <p>This event is fired on the {@link MinecraftForge#EVENT_BUS main Forge event bus}</p>
```
Some suggestions to improve the javadoc. Please also confirm if the event is fired on both logical sides or if only fired on the logical server. (My javadoc suggestion assumes that it is fired only on the logical server.)

```suggestion
 * Fired to check if the given item is currently allowed to be picked up or not.
 *
 * <p>This event is not {@linkplain Cancelable cancellable}, and {@linkplain HasResult has a result}.</p>
 * <ul>
 *     <li>{@link Result#ALLOW} - Forcibly allow this item to be picked up.</li>
 *     <li>{@link Result#DEFAULT} - Let the default logic dictate if this item should be picked up or not. This relies on the {@link ItemEntity#isPickupAllowed(Object, ItemPickupReason...)} method.</li>
 *     <li>{@link Result#DENY} - Forcibly disallow this item to be picked up.</li>
 * </ul>
 *
 * <p>This event is fired on the {@linkplain MinecraftForge#EVENT_BUS main Forge event bus},
 * only on the {@linkplain LogicalSide#SERVER logical server}.</p>
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/item/ItemPickupAllowedEvent.java:
```diff
@@ -0,0 +1,69 @@
+        return getEntity().blockPosition();
+    }
+
+    public Object getCollector()
+    {
+        return collector;
+    }
```
It is unclear what possible types this may be, at least for the built-in item pickup reasons. At the very least, the javadocs for those built-in reasons should mention what the type of the collector is.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Class containing all standard ItemPickupReasons provided by Forge.
+ *
+ * @see ItemPickupReasons
```
This is a self-referential `@see` tag. Perhaps you were aiming for the `ItemPickupReason` class?

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @see ItemPickupReasons
+ */
+public final class ItemPickupReasons
```
Utility classes like these should have a private constructor to prevent instantiation.

sciwhiz12 commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -276,6 +276,7 @@ modId="data_pack_registries_test"
 modId="creative_mode_tab_test"
 [[mods]]
 modId="trade_with_villager_event_test"
-
```
Please avoid removing blank lines in here.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -0,0 +1,125 @@
+import java.util.List;
+
+@Mod(ItemPreventPickupTest.ID)
+public class ItemPreventPickupTest
```
Please provide a javadoc for this class which explains what feature this tests, how to test it, and what the expected results are from said test.

Additionally, please fix the formatting in this class: there should be spaces between the condition and the keyword in an `if` expression (`if (...)`)

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -0,0 +1,125 @@
+import net.minecraft.tags.BlockTags;
+import net.minecraft.world.entity.monster.Zombie;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.*;
```
Avoid wildcard imports.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -0,0 +1,125 @@
+    private void onItemPickupAllowed_Pre(ItemPickupAllowedEvent event)
+    {
+        // only for new dummy item, and only add the queries once per item entity
+        if(ENABLE && event.getItemStack().is(DUMMY.get()) && !event.getEntity().getPersistentData().getBoolean("ForgeInitItemPickupQuery"))
```
Is this really how pickup queries are designed to be added to item entities, by specifying a high priority event handler and using persistent data to track if the query has been added? It seems a bit... clunky to me, at least with the use of persistent data.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -0,0 +1,125 @@
+        if(ENABLE && event.getItemStack().is(DUMMY.get()) && !event.getEntity().getPersistentData().getBoolean("ForgeInitItemPickupQuery"))
+        {
+            // hoppers only pickup if item is stacked more than 1
+            if(event.hasPickupReason(ItemPickupReasons.HOPPER)) event.getEntity().addPickupQuery(IF_MORE_THAN_ONE);
```
Shouldn't the pickup reason be tested inside the query itself, not during the event listener that adds the pickup query? What happens if a pickup is attempted by a non-hopper and gets rejected (which still triggers this code), and later a pickup is attempted by a hopper?

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -0,0 +1,95 @@
+    @Override
+    public boolean equals(Object obj)
+    {
+        if(this == obj) return true;
+        if(obj instanceof ItemPickupReason other) return name.equals(other.name);
```
Formatting: spaces between `if` keyword and opening parenthesis.
```suggestion
        if (this == obj) return true;
        if (obj instanceof ItemPickupReason other) return name.equals(other.name);
```

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -0,0 +1,95 @@
+     *
+     * @return All registered pickup reasons.
+     */
+    public static Collection<ItemPickupReason> getPickupReasons()
```
Could also be named something like `all()`, so it is called as `ItemPickupReason.all()` instead of `ItemPickupReason.getPickupReasons()`. Another suggestion would be to move this to `ItemPickupReasons`.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -0,0 +1,95 @@
+     * @param name Name of the pickup reason.
+     * @return Gets or creates a given pickup reason.
+     */
+    public static ItemPickupReason getOrCreate(String name)
```
Should we suggest that mods add their mod ID as a namespaced prefix to their pickup reasons?

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -0,0 +1,125 @@
+        if(ENABLE)
+        {
+            // forcibly disallow picking up iron ingots
+            if(event.getItemStack().is(Tags.Items.INGOTS_IRON)) event.setResult(Event.Result.DENY);
```
Formatting: for `if` statements with `else` clauses, please move all code statements to new lines instead of being inline with the `if`/`else`.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupQuery.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    boolean query(@NotNull ItemEntity item, @NotNull ItemStack stack, @NotNull Level level, @NotNull BlockPos pos, @NotNull Object collector, @NotNull ItemPickupReason... pickupReasons);
+
+    // vanilla Java Util Predicate-like methods
```
How useful are these methods? If there's little chance of them being used, I see no reason to include them.

sciwhiz12 commented on patches/minecraft/net/minecraft/world/entity/item/ItemEntity.java.patch:
```diff
@@ -144,3 +145,88 @@
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
++      var eventResult = event.getResult();
++
++      if(eventResult == net.minecraftforge.eventbus.api.Event.Result.ALLOW) return true;
```
Formatting: space between `if` keyword and opening parenthesis. (Same comment for the rest of this method.)

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupQuery.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Objects;
+import java.util.function.Predicate;
+
+/**
+ * FunctionalInterface used to query the Level to determine if the specified ItemEntity is valid to be picked up or not.
+ */
+@FunctionalInterface
+public interface ItemPickupQuery
```
The javadoc should clarify what "valid to be picked up or not" actually means. What does returning `true` or `false` in the interface's method mean for the pickup logic?

ApexModder committed:
Fix issues with pr and implement suggestions.

Added & improved various java-docs.
Added missing white space around 'if's
Add new line back to mods.toml
Changed ItemPickupQuery to take in a Set of ItemPickupReason rather than a varargs array
Added ApiStatus annotations

Renamed ItemPickupReason.getPickupReasons to ItemPickupReason.all
Moved ItemPickupReason.all and ItemPickupReason.getOrCreate into ItemPickupReasons

Merged Hopper test query registration into main onItemPickupAllowed event listener
Using a higher priority & persistent data is not required
The backing collection is a Set, the query will only get added to it once
Fixed for matting issues in ItemPreventPickupTest
Moved Hopper query logic from ItemPickupAllowedEvent listener into the query itself

ApexModder committed:
Fix formatting issue in ItemPickupReason

ApexModder committed:
Replace wildcard imports with fully qualified imports

Matyrobbrt commented on patches/minecraft/net/minecraft/world/entity/Mob.java.patch:
```diff
@@ -24,6 +24,14 @@
+    }
+ 
+    protected void m_7581_(ItemEntity p_21471_) {
++      if(!p_21471_.isPickupAllowed(this, net.minecraftforge.items.pickup.ItemPickupReasons.LIVING_ENTITY)) return; // Forge: Check to see if item currently disallows being picked up for this entity
```
Formatting: space between `if` and condition. Additionally, change the comment to:
```diff
-Check to see if item currently disallows being picked up for this entity
+Check to see if the item currently disallows being picked up by this entity
```

Matyrobbrt commented on patches/minecraft/net/minecraft/world/entity/item/ItemEntity.java.patch:
```diff
@@ -104,11 +104,12 @@
     public void m_6123_(Player p_32040_) {
        if (!this.f_19853_.f_46443_) {
 +         if (this.f_31986_ > 0) return;
++         if(!isPickupAllowed(p_32040_, net.minecraftforge.items.pickup.ItemPickupReasons.PLAYER)) return; // Forge: Check to see if item currently disallows being picked up by players
```
Formatting: space between `if` and the condition. Additionally, add a `the` between `if` and `item` in the comment.

Matyrobbrt commented on patches/minecraft/net/minecraft/world/entity/item/ItemEntity.java.patch:
```diff
@@ -144,3 +145,89 @@
++   public final void removePickupQuery(net.minecraftforge.items.pickup.ItemPickupQuery pickupQuery)
++   {
++      pickupQueries.remove(pickupQuery);
++   }
++
++   /**
++    * Returns true if this item is allowed to be picked up for the given {@link net.minecraftforge.items.pickup.ItemPickupReason reasons}
++    *
++    * @param collector The object attempting to pick up this item
++    * @param pickupReasons The reasons this item is being picked up
++    * @return if this item is allowed to be picked up for the given {@link net.minecraftforge.items.pickup.ItemPickupReason reasons}
++    */
++   public final boolean isPickupAllowed(Object collector, net.minecraftforge.items.pickup.ItemPickupReason... pickupReasons)
++   {
++      // check forge event to see if item is forcibly allowed to be picked up or not
++      var pickupReasonSet = com.google.common.collect.ImmutableSet.copyOf(pickupReasons);
++      var event = new net.minecraftforge.event.entity.item.ItemPickupAllowedEvent(this, collector, pickupReasonSet);
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
++      var eventResult = event.getResult();
++
++      if (eventResult == net.minecraftforge.eventbus.api.Event.Result.ALLOW) return true;
++      if (eventResult == net.minecraftforge.eventbus.api.Event.Result.DENY) return false;
++
++      // check if vanilla flag has been set, if disallow pickup
++      if (lifespan == f_149661_) return false;
++
++      // check to see if any specified pickup reason is disallowed
++      for (var pickupReason : pickupReasons)
++      {
++         if (disabledPickupReasons.contains(pickupReason)) return false;
++      }
++
++      // check to see if any query currently disallows picking up item
++      for (var pickupQuery : pickupQueries)
++      {
++         if (!pickupQuery.query(this, m_32055_(), f_19853_, m_20183_(), collector, pickupReasonSet)) return false;
++      }
++
++      return true;
++   }
```
Patches use same-line curly brackets.

Matyrobbrt commented on patches/minecraft/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch:
```diff
@@ -28,6 +28,15 @@
+ 
+    public static boolean m_59331_(Container p_59332_, ItemEntity p_59333_) {
++      // Forge: Passing the container as the collector instance, as it is the Hopper [BlockEntity/MinecartEntity] instance
++      if(!p_59333_.isPickupAllowed(p_59332_, net.minecraftforge.items.pickup.ItemPickupReasons.HOPPER)) return false; // Forge: Check to see if item currently disallows being picked up by hoppers
```
Formatting: space between `if` and condition.

Matyrobbrt commented on src/main/java/net/minecraftforge/event/entity/item/ItemPickupAllowedEvent.java:
```diff
@@ -0,0 +1,77 @@
+        this.pickupReasons = pickupReasons;
+    }
+
+    public ItemStack getItemStack()
+    {
+        return getEntity().getItem();
+    }
+
+    public Level getLevel()
+    {
+        return getEntity().getLevel();
+    }
+
+    public BlockPos getPos()
+    {
+        return getEntity().blockPosition();
+    }
```
These methods seem unnecessary utility methods which Forge avoids to add.

Matyrobbrt commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Class representing a reason in which an ItemEntity is being picked up.
+ */
+public final class ItemPickupReason implements StringRepresentable, Comparable<ItemPickupReason>
```
Why exactly is this comparable?

ApexModder replied:
No reason really, tbh don't know why I originally made it comparable, I will remove the Comparable interface.

Matyrobbrt commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupQuery.java:
```diff
@@ -0,0 +1,39 @@
+ * <p>All queries must return true, in order for the item to be picked up by the collector.
+ */
+@FunctionalInterface
+public interface ItemPickupQuery
```
This interface is better named `ItemPickupPredicate`

ApexModder committed:
Rename ItemPickupQuery to ItemPickupPredicate
Fix formatting issue in patch files
Remove unwanted utility methods from ItemPickupAllowedEvent
Remove Comparable interface from ItemPickupReason
Fix issues in ItemPreventPickupTest caused by the above changes

ApexModder committed:
Update comment in Mob

Matyrobbrt commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupPredicate.java:
```diff
@@ -0,0 +1,37 @@
+     * @param collector Object trying to pick up the ItemEntity or null
+     * @return if the specified ItemEntity is valid to be picked up
+     */
+    boolean query(@NotNull ItemEntity item, @NotNull ItemStack stack, @NotNull Level level, @NotNull BlockPos pos, @NotNull Object collector, @NotNull Set<ItemPickupReason> pickupReasons);
```
Please rename this method to `test` and all references to this interface from `Query` -> `Predicate` in methods or docs

ApexModder committed:
Rename ItemPickupPredicate.query to ItemPickupPredicate.test
Rename all references of ItemPickupPredicate from 'query' to 'predicate'

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/item/ItemPickupAllowedEvent.java:
```diff
@@ -6,6 +6,7 @@
 package net.minecraftforge.event.entity.item;
 
 import org.apache.commons.lang3.ArrayUtils;
```
Since `ArrayUtils` is no longer used in this class, this import should be removed.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -18,13 +19,24 @@
+     * package level of visibility to be accessible in {@link ItemPickupReasons}.
+     * <p>Modders should <b>NOT</b> access this field.
+     */
+    @ApiStatus.Internal
```
I'm unsure if we need to mark package-private fields and methods with `@ApiStatus.Internal`. Until there is a discussion and decision otherwise, these can remain here.

However, the comment should likely be downgraded from a javadoc comment to a regular comment. It additionally does not need to be marked by the comment itself as being internal to Forge, as the annotation already signals that.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -18,13 +19,24 @@
+     * @implNote Invoking this directly will <b>not</b> register the reason correctly to the internal reasons map.
+     */
+    @ApiStatus.Internal
+    ItemPickupReason(String name)
```
Same remarks as I had with the field in this class.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -71,25 +77,4 @@ public String toString()
-    public static ItemPickupReason getOrCreate(String name)
-    {
-        return reasons.computeIfAbsent(name, ItemPickupReason::new);
-    }
```
Since these were moved to `ItemPickupReasons` the utility class, the class javadoc should at the very least add a `@see` to that class.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -5,21 +5,70 @@
  */
-public final class ItemPickupReasons
+@ApiStatus.NonExtendable
+public interface ItemPickupReasons
```
I'm reluctant with the use of an interface for this utility class, even if marked with `@ApiStatus.NonExtendable`. I don't think using an interface for this purpose -- being a holder of constants that will never be implemented itself -- is appropriate. This should just be a regular class, with a private constructor and all its methods and fields being marked as `public static`.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -5,21 +5,70 @@
-    public static final ItemPickupReason PLAYER = getOrCreate("player");
+    /**
+     * Player is attempting to pick up an item
+     * <p>{@code collector} types <i>should</i> result in the {@link Player} attempting to pick up the item.
```
A bit of a nitpick, but I don't see an reason why "should" is emphasized.

ApexModder replied:
I emphasized the `should` as the collector should always be the player picking up the item for `ItemPickupReasons#PLAYER` but because it is an `Object` type and relies on people passing the correct collector when invoking `ItemEntity#isPickupAllowed` it could technically resolve to something other than a player if people misuse the system.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -5,21 +5,70 @@
+    /**
+     * Returns all registered pickup reasons.
+     *
+     * @return All registered pickup reasons.
```
This should use the inline `{@return}` tag.

```suggestion
     * {@return all registered pickup reasons}
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -20,22 +24,46 @@
 public class ItemPreventPickupTest
 {
-    public static final boolean ENABLE = false;
+    public static final boolean ENABLE = true;
```
Given this test's effects on the pickupability of diamonds and iron ingots, this should be disabled by default.

ApexModder replied:
Whoops, how did I leave that as true, swore I changed that back to false prior to committing but clearly not, I will change that back now.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/ItemPreventPickupTest.java:
```diff
@@ -20,22 +24,46 @@
+ * in which should only be picked up under the following circumstances
+ * <ul>
+ *     <li>If Player is sneaking</li>
+ *     <li>If Zombie is standing on a Stone block</li>
```
It should be noted that the Zombie needs to have the capacity to pickup loot, as controlled by the `CanPickUpLoot` byte tag, with a 55% chance of being randomly enabled in spawned zombies. A zombie which can pickup loot is spawned by the following command: `/summon minecraft:zombie ~ ~ ~ {CanPickUpLoot:1b}`.

sciwhiz12 reviewed (changes requested):
This PR works fine from my testing.

Please rebase your PR on the latest main branch, so the compatibility checks succeed.

---

I do wish to give concern to a thought on possible performance with regards to many item entities laying on the ground (such as those which may happen for large-scale farms that don't have a proper working output) and a player or multiple coming near enough to trigger the pickup mechanic. That would cause a lot of events to get fired, one per collector-item entity pair.

However, I don't think that is reason to deny this PR, since that will likely be inherent in any system we design for pickup checks between item entities and their collectors.

ApexModder committed:
Optimize imports

ApexModder committed:
Fix issues with java-docs

ApexModder committed:
Disable ItemPreventPickupTest

ApexModder committed:
Add note about zombies needing a special tag to always pickup loot in java-docs

ApexModder committed:
Convert ItemPickupReasons from interface to utility-class

ApexModder committed:
Remove unnecessary emphasis from ItemPickupReasons java-docs

ApexModder committed:
Fix issue in java-doc caused by changing ItemPickupReasons from interface to class

ApexModder commented:
> Please rebase your PR on the latest main branch, so the compatibility checks succeed.

Uhh did I rebase wrong? why did a ton of commits appear here?
I used IntelliJ "sync-fork" which did a rebase with upstream afaik, was that wrong?

sciwhiz12 commented:
You did not properly rebase your PR on the main branch, as evidenced by the GitHub commit log showing many commits from the main branch being committed under your name and showing after your own commits.

You will have to revert the rebase and either merge the main branch using `git merge`, or rebase your PR branch on the main branch using `git rebase` and force-push to your remote PR branch with `git push --force-push`.

---

You can revert the rebase by doing `git reset --hard 861be0fd462f95d854930eef5211c3b56e41d926` while your PR branch is checked out (`git switch feature/1.19.x-item-magnet-flag`), which will reset the branch to the commit before your ill-fated rebase (https://github.com/MinecraftForge/MinecraftForge/pull/9112/commits/861be0fd462f95d854930eef5211c3b56e41d926).

After reverting the rebase and ensuring that your local main branch is up-to-date with the upstream main branch, you may either:
- Merge the main branch using `git merge 1.19.x`, and pushing the resulting merge commit; or
- Rebase your PR branch on the main branch using `git rebase 1.19.x`, and force-pushing the resulting rebased commits to your fork using `git push --force-with-lease 1.19.x-item-magnet-flag`. (See [this StackOverflow answer](https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force) on the difference between `--force` and `--force-with-lease`).

Note that rebasing can be more difficult depending on the amount of commits on your PR branch and the amount of conflicts between your PR branch and the main branch. Because rebasing entails re-applying changes of your commits on top of the main branch, if the main branch modifies a patch which you touch, each commit that touches that patch will require manual conflict resolution. 

However, rebasing leaves a _linear_ history, rather than the converging history from a merge commit, which may be more desirable as a cleaner commit history in your PR.

_Copied and adapted from https://github.com/MinecraftForge/MinecraftForge/pull/9269#issuecomment-1407423289_

ApexModder commented:
Hopefully, I applied the rebase correctly this time
Thanks for the explanation @sciwhiz12 :)

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -0,0 +1,75 @@
+ * <p>Make use of {@link #getOrCreate(String)} for building new {@link ItemPickupReason} instances.
+ */
+@ApiStatus.NonExtendable
+public class ItemPickupReasons
```
I forgot to note that the class should be `final`.

sciwhiz12 commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -0,0 +1,75 @@
+        return ItemPickupReason.reasons.computeIfAbsent(name, ItemPickupReason::new);
+    }
+
+    private ItemPickupReasons() {
```
Formatting: braces on newlines.
```suggestion
    private ItemPickupReasons()
    {
```

sciwhiz12 reviewed (changes requested):
Just a few minor things left to take care of.

ApexModder committed:
Mark ItemPickupReasons as being final
Fix formatting issue

LexManos commented on patches/minecraft/net/minecraft/world/entity/item/ItemEntity.java.patch:
```diff
@@ -144,3 +145,82 @@
++
++   // FORGE START
++   private final java.util.Set<net.minecraftforge.items.pickup.ItemPickupReason> disabledPickupReasons = com.google.common.collect.Sets.newHashSet();
++   private final java.util.Set<net.minecraftforge.items.pickup.ItemPickupPredicate> pickupPredicates = com.google.common.collect.Sets.newHashSet();
```
Now that we are past java 7. We use the diamond operator instead of google's helpers.

LexManos commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupPredicate.java:
```diff
@@ -0,0 +1,37 @@
+     * @param item The ItemEntity being picked up
+     * @param stack The ItemStack being picked up
+     * @param level Level containing the ItemEntity
+     * @param pos Position of the ItemEntity
```
The item stack, level, and position are all redundant as the Item entity contains getters for all of these. Allowing them to be passed in allows for the opportunity for disagreements.

LexManos commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReason.java:
```diff
@@ -0,0 +1,77 @@
+ * @see ItemPickupReasons
+ */
+public final class ItemPickupReason implements StringRepresentable
+{
```
Couldn't this entire class be replaced with `record ItemPickupReason(String name){}`
Records implement the equals/hashcodes/ for us.
Tho we would loose the instance equality because there is no way to make the constructor private. 
However we are using them in sets, which use hashCode/equals anyway, so there is no performance difference.
Also, why are you implementing StringRepresentable?

ApexModder replied:
> Also, why are you implementing StringRepresentable?

tbh I have gotten into a bit of a bad habit with this interface, if an object has a String property representing its internal/serialized name, I tend to implement that interface without thinking, usually though it's for enums/block states which makes some sense there.

LexManos commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -0,0 +1,76 @@
+     * Gets a given pickup reason or creates a new instance of none exists.
+     * <p>For custom modded reasons, it is recommended you prefix the name with your mod id
+     * <i>[{@code {modId}:{reasonName}}]</i>, this is to prevent registering custom reasons
+     * with duplicate names, and potentially using custom mod reasons in places you did not intend on using them.
```
If we're telling people to prefix, then just accept a ResourceLocation

ApexModder replied:
to be honest, I personally don't like methods that take in a ResourceLocation, I would rather just pass in a String, it looks cleaner to me than creating a ResourceLocation.

`ItemPickupREasons.getOrCreate("player");`

vs

`ItemPickupReasons.getOrCreate(new ResourceLocation("forge:player"));`
`ItemPickupReasons.getOrCreate(new ResourceLocation("forge", "player"));`

could change the method to take in 2 Strings, `mod id` & `name` which would combine the 2 into a prefixed String or wrap directly into a ResourceLocation instance.

`ItemPickupReasons.getOrCreate("forge", "player");`

LexManos commented on src/main/java/net/minecraftforge/items/pickup/ItemPickupReasons.java:
```diff
@@ -0,0 +1,76 @@
+     */
+    public static Collection<ItemPickupReason> all()
+    {
+        return Collections.unmodifiableCollection(ItemPickupReason.reasons.values());
```
Why would anyone need access to all values?

ApexModder replied:
people shouldn't need to access all values but I initially copied `ToolAction` which too has a method to obtain all registered values, could simply remove this method.
`ToolAction#getActions`

This also goes for `ItemPickupReason` not being a record, while it could be a record, I initially copied `ToolAction` which is not a record.
[_but now that I think of it, that may be because it was created before records were a thing?_]

LexManos reviewed (commented):
Overall I am not a fan of this change, the implementation is bad IMO as this context object being passed in as a raw object is just asking for casting errors. We also already have 3 hooks for controlling when players can pickup items. In the playerTouch function itself.. So we should probably deprecate or remove those events so that we don't have redundancy.

ApexModder committed:
Deprecate old event used to allow modders to dictate how/when players can pickup items.

ApexModder committed:
Remove params from ItemPickupPredicate which can easily be obtained from the passed ItemEntity
This is to fix potential disagreements, passing ItemEntity with differing BlockPos/Level/ItemStack than what was passed in.

ApexModder commented:
> Overall I am not a fan of this change, the implementation is bad IMO as this context object being passed in as a raw object is just asking for casting errors. We also already have 3 hooks for controlling when players can pickup items. In the playerTouch function itself.. So we should probably deprecate or remove those events so that we don't have redundancy.

correct me if I'm wrong but the existing stuff only allows changing when/how player entities can/should be allowed to pickup items. there's nothing more generalized to dictate how "anything" can/should pickup items? ("anything" being blocks, entities, or custom systems as a whole, magnets).

also, I didn't know a better way of passing the context of who/what is trying to pickup the item than to pass a raw object type, sure it's not the best and may potentially cause casting issues but how else would you have a param that can expect "any" type? (entities, block entities, or some custom modded type).

ApexModder committed:
Swap to use diamond operator rather than googles utility methods

ApexModder committed:
Fix typo in EntityItemPickupEvent java-doc leading to incorrect class

ApexModder committed:
Stop implementing StringRepresentable interface on ItemPickupReason

ApexModder committed:
Merge branch '1.19.x' into 1.19.x-item-magnet-flag
